colexec: fixed nullableArgs projection operators

wenyihu6 opened the pull request:
Previously, the `Concat` projection operators on arrays directly copy nulls into
the output without performing the actual projection whenever either argument was
null. This is incorrect because `Concat` operator can actually handle null arguments,
and we want the actual projection result instead. Since currently the only
projection operator that needs `nullableArgs == true` behaviour is
`ConcatDatumDatum`, this PR fixed this by adding a boolean field `nullableArgs`
to the logic for operators with `Datum`, `Datum`. If we later introduce another
projection operation that has nullableArgs == true, we should update this code
accordingly.

Fixes: https://github.com/cockroachdb/cockroach/issues/83091

Release note (bug fix): Fixed a bug in `Concat` projection operators on arrays
that gave output of nulls when the projection operator can actually handle null
arguments and may result in a non-null output.

wenyihu6 committed:
colexec: fixed nullableArgs projection operators

Previously, the `Concat` projection operators on arrays directly copy nulls into
the output without performing the actual projection whenever either argument was
null. This is incorrect because `Concat` operator can actually handle null arguments,
and we want the actual projection result instead. Since currently the only
projection operator that needs `nullableArgs == true` behaviour is
`ConcatDatumDatum`, this PR fixed this by adding a boolean field `nullableArgs`
to the logic for operators with `Datum`, `Datum`. If we later introduce another
projection operation that has nullableArgs == true, we should update this code
accordingly.

Fixes: https://github.com/cockroachdb/cockroach/issues/83091

Release note (bug fix): Fixed a bug in `Concat` projection operators on arrays
that gave output of nulls when the projection operator can actually handle null
arguments and may result in a non-null output.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/84159)
<!-- Reviewable:end -->

mgartner reviewed (commented):
Great find!

I wonder if the `if nullableArgs { .. }` should be conditionals in the template performed at "compile time" rather than run time. Maybe @yuzefovich can chime in on that.

> Reviewed 3 of 3 files at r1, 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-:-N6hMh-M6PKEPlvcojTe:b-fudrt1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft, @wenyihu6, and @yuzefovich)

___
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 119 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6hMu7F6hu2BWOA0r2s:-N6hMu7GCl_UW4gvSRtR:br5eul2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cf421c6a91e435ba98401218f5725a85786c27c/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L119)):*
> ```Go
> 		col1 := vec1._L_TYP()
> 		col2 := vec2._R_TYP()
> 		nullableArgs := p.nullableArgs
> ```

nit: this looks unused, should it be removed?
___
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6hMFrMDOW-hfaEEGql:-N6hMFrMDOW-hfaEEGqm:bzb0wmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5e841324c8d16047097e98afa4f04fff1b5e2a9/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L163)):*
> ```Go
> 	//(1) _HAS_NULLS is true:
> 	//If nullableArgs is false, union _outNulls with the set of input Nulls.
> 	//If nullableArgs is true and nullableArgs is true, the function’s definition
> ```

nit: "nullableArgs is true" is repeated


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 119 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6hMu7F6hu2BWOA0r2s:-N6hT6hK3iOfDhyeYGbb:b-8sl9p8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cf421c6a91e435ba98401218f5725a85786c27c/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L119)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: this looks unused, should it be removed?
</blockquote></details>

It is defined here for `_SET_PROJECTION` and `_SET_SINGLE_TUPLE_PROJECTION` to reference later. Alternatively, we could also directly reference `p.nullableArgs` inside the functions. I defined them here because we are defining a bunch of variables that are also used later here e.g.`vec1`, `vec2`, `col1`, `col2`.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
Yeah. Yahor mentioned to me that there will be other ways to improve the performance, and this pr will require some iterations. 





<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Nice!

I think this already fixes the bug, but I'd like to make sure that performance doesn't suffer because of the additional condition. We should run the microbenchmarks to see whether there is any impact, and if there isn't, then we're good to go :)

If there is some slowdown, we have two ways for how to address it:
- the way that Marcus suggested - for each projection operation we'd create two structs (one for `nullableArgs==false` and another for `nullableArgs==true` cases), then in `GetProjectionOperator` we'd be able to pick the correct one. However, this results in doubling the generated code which is not great. Especially so because only very few operators have `nullableArgs==true` case, meaning we'd be generating a lot of dead code.
- another way is to take advantage of the fact that very few operators need this `nullableArgs==true` behavior, so we can special case them. In other words, we'd keep the code for all `nullableArgs==false` cases unchanged and would modify the template slightly for a few operators (I think it would actually be just one - `ConcatDatumDatum`). This would be the cleanest option overall, but it is slightly "unsafe" in a sense that if we later introduce another projection operator with `nullableArgs==true`, we might forget to update the vectorized code accordingly. However, I think this would still be my preference given that we introduce new operators very rarely nowadays.

We'll explore these options with Wenyi during a pair programming session.

> Reviewed 10 of 10 files at r3, 8 of 8 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-:-N6hygWHE1DYH17lcBk7:bm3d30y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rytaft, and @wenyihu6)

___
*[`-- commits` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6hxCcYBxHVuEE0Ivb7):*
nit: we usually just use package names, i.e. `colexec`, without the `1-` part.
___
*[`-- commits` line 5 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6hxMfYE1Yp3_YBU1WL):*
nit: missing release note.
___
*[`-- commits` line 8 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6i-ogCBsfLU_A2RCKw):*
nit: I'd probably squash both commits into one since it's a single logical change.
___
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 159 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6hxV_h91XvOtDzlfQG:-N6hxV_h91XvOtDzlfQH:bfmwuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b491c2f6bce27e5f6ff8fef0db134377680c4971/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L159)):*
> ```Go
> 		}
> 	}
> 	//_outNulls has been updated from within the _ASSIGN function to include any
> ```

nit: we usually add a space after `//` in comments.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`-- commits` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6hxCcYBxHVuEE0Ivb7):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we usually just use package names, i.e. `colexec`, without the `1-` part.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 159 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6hxV_h91XvOtDzlfQG:-N6i9_gE24ehNsiThNx0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b491c2f6bce27e5f6ff8fef0db134377680c4971/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L159)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we usually add a space after `//` in comments.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner commented:
> the way that Marcus suggested - for each projection operation we'd create two structs (one for nullableArgs==false and another for nullableArgs==true cases), then in GetProjectionOperator we'd be able to pick the correct one. However, this results in doubling the generated code which is not great. Especially so because only very few operators have nullableArgs==true case, meaning we'd be generating a lot of dead code.

@yuzefovich Since `nullableArgs` is a property of each operator, wouldn't we only need a single version of each - compile with either the functionality when `nullableArgs=true` or when `nullableArgs=false`?

yuzefovich reviewed (commented):
> Since `nullableArgs` is a property of each operator, wouldn't we only need a single version of each - compile with either the functionality when `nullableArgs=true` or when `nullableArgs=false`?

That is true, but that would be extremely hard to actually plumb into the template engine (since it is completely unaware of different properties of projection operators, and the templating for projections is already super-complicated).

----

We actually ran the benchmarks with Wenyi, and they showed no significant changes (minor slowdown, but the throughput is on the order of GBs/s, so it doesn't really matter). She also had a clever idea of how to push the `if` conditional out of the tight inner loop and do it only once per batch which seems good to me.

> Reviewed 6 of 6 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-:-N6ipZAEAZpV1Tw5CBBF:bx0m32w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rytaft, and @wenyihu6)

___
*[`-- commits` line 11 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6iqBI-7Wmeg1CQobNw):*
I think simply adding the test case from Becca's repro is sufficient.
___
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 119 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6hMu7F6hu2BWOA0r2s:-N6iqiVn-iEogMabwUEg:b-w01da8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cf421c6a91e435ba98401218f5725a85786c27c/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L119)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

It is defined here for `_SET_PROJECTION` and `_SET_SINGLE_TUPLE_PROJECTION` to reference later. Alternatively, we could also directly reference `p.nullableArgs` inside the functions. I defined them here because we are defining a bunch of variables that are also used later here e.g.`vec1`, `vec2`, `col1`, `col2`.
</blockquote></details>

I think we can now remove the local variable and use `p.nullableArgs` directly since we no longer access it in `_SET_PROJECTION` and `_SET_SINGLE_TUPLE_PROJECTION` template functions.
___
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6hMFrMDOW-hfaEEGql:-N6ir8DJ90C9p0oXvoue:bo7lmp5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5e841324c8d16047097e98afa4f04fff1b5e2a9/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L163)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

</details>
Yeah. Yahor mentioned to me that there will be other ways to improve the performance, and this pr will require some iterations.

</blockquote></details>

nit: I'd keep the comment here unchanged and would move the discussion about `nullableArgs` to where it is used in `Next`.
___
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 170 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6irG0g3fxwRUwBYvYq:-N6irG0hDMGBghvj4Yiu:b-kq5ghf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4915731e50c9e8f9467f09f03d4941c4a099e581/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L170)):*
> ```Go
> 	// {{if _HAS_NULLS}}
> 	projVec.SetNulls(_outNulls.Or(*col1Nulls).Or(*col2Nulls))
> 	
> ```

nit: new empty line.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
This new commit changed the if statement to use nullableArgs only for functions
[`ConcatDatumDatum`](https://github.com/cockroachdb/cockroach/blob/7d7facf1eb091b8dbe085437a9f2fe9f486b397d/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L125-L136). I also changed the logic a bit. Instead of checking for
`nullableArgs` whenever `_HAS_NULLS` is true in a if statement, I changed the
logic to pass in `_HAS_NULLS` as false when `nullableArgs` is true. In other
words, we are treating `_HAS_NULLS` as false if `nullableArgs` is true. When
`nullableArgs` is true, the projection operator is capable of handling nullable
argument. And we don't need to copy null values into the output as well. So this
logic should be equivalent to what I had previously. This is the new benchmark
result (see below), and the performance changes is smaller than what I had before.

This new commit also added some test cases under `pkg/sql/logictest/testdata/logic_test/array`.

```
wenyihu@crlMBP-C02ZV00XMD6TMjA5 benchmark % ./benchstat old.txt new.txt
name                                                            old time/op    new time/op    delta
ProjOp/projPlusInt64Int64/useSel=false/hasNulls=false-16          1.13µs ± 0%    1.40µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64Const/useSel=false/hasNulls=false-16     1.00µs ± 0%    1.04µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64/useSel=false/hasNulls=true-16           2.70µs ± 0%    3.05µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64Const/useSel=false/hasNulls=true-16      1.92µs ± 0%    2.36µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64/useSel=true/hasNulls=false-16           2.13µs ± 0%    2.88µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64Const/useSel=true/hasNulls=false-16      1.79µs ± 0%    2.37µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64/useSel=true/hasNulls=true-16            3.73µs ± 0%    5.15µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64Const/useSel=true/hasNulls=true-16       2.74µs ± 0%    4.26µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64/useSel=false/hasNulls=false-16         1.04µs ± 0%    1.59µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64Const/useSel=false/hasNulls=false-16     898ns ± 0%    1144ns ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64/useSel=false/hasNulls=true-16          2.49µs ± 0%    3.54µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64Const/useSel=false/hasNulls=true-16     1.95µs ± 0%    2.77µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64/useSel=true/hasNulls=false-16          1.89µs ± 0%    2.68µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64Const/useSel=true/hasNulls=false-16     1.64µs ± 0%    2.21µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64/useSel=true/hasNulls=true-16           3.30µs ± 0%    4.14µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64Const/useSel=true/hasNulls=true-16      2.60µs ± 0%    3.39µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64/useSel=false/hasNulls=false-16          1.40µs ± 0%    1.79µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64Const/useSel=false/hasNulls=false-16     1.11µs ± 0%    1.40µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64/useSel=false/hasNulls=true-16           2.73µs ± 0%    3.49µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64Const/useSel=false/hasNulls=true-16      2.32µs ± 0%    2.86µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64/useSel=true/hasNulls=false-16           2.48µs ± 0%    3.00µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64Const/useSel=true/hasNulls=false-16      2.01µs ± 0%    2.46µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64/useSel=true/hasNulls=true-16            3.56µs ± 0%    4.54µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64Const/useSel=true/hasNulls=true-16       2.82µs ± 0%    3.54µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64/useSel=false/hasNulls=false-16            171µs ± 0%     208µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64Const/useSel=false/hasNulls=false-16       130µs ± 0%     167µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64/useSel=false/hasNulls=true-16             137µs ± 0%     197µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64Const/useSel=false/hasNulls=true-16        130µs ± 0%     188µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64/useSel=true/hasNulls=false-16             191µs ± 0%     235µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64Const/useSel=true/hasNulls=false-16        149µs ± 0%     190µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64/useSel=true/hasNulls=true-16              130µs ± 0%     213µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64Const/useSel=true/hasNulls=true-16         138µs ± 0%     173µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64/useSel=false/hasNulls=false-16            1.05µs ± 0%    1.50µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64Const/useSel=false/hasNulls=false-16        868ns ± 0%    1179ns ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64/useSel=false/hasNulls=true-16             2.28µs ± 0%    3.34µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64Const/useSel=false/hasNulls=true-16        1.63µs ± 0%    2.18µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64/useSel=true/hasNulls=false-16             3.19µs ± 0%    4.88µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64Const/useSel=true/hasNulls=false-16        2.44µs ± 0%    4.20µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64/useSel=true/hasNulls=true-16              3.03µs ± 0%    4.02µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64Const/useSel=true/hasNulls=true-16         2.33µs ± 0%    3.16µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64/useSel=false/hasNulls=false-16            1.01µs ± 0%    1.35µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64Const/useSel=false/hasNulls=false-16       1.00µs ± 0%    1.12µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64/useSel=false/hasNulls=true-16             2.37µs ± 0%    3.10µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64Const/useSel=false/hasNulls=true-16        1.80µs ± 0%    2.19µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64/useSel=true/hasNulls=false-16             3.42µs ± 0%    4.98µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64Const/useSel=true/hasNulls=false-16        3.06µs ± 0%    4.40µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64/useSel=true/hasNulls=true-16              3.34µs ± 0%    4.13µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64Const/useSel=true/hasNulls=true-16         2.33µs ± 0%    3.11µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64/useSel=false/hasNulls=false-16            1.02µs ± 0%    1.40µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64Const/useSel=false/hasNulls=false-16        862ns ± 0%    1248ns ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64/useSel=false/hasNulls=true-16             2.33µs ± 0%    3.22µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64Const/useSel=false/hasNulls=true-16        1.66µs ± 0%    2.16µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64/useSel=true/hasNulls=false-16             3.18µs ± 0%    4.63µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64Const/useSel=true/hasNulls=false-16        2.34µs ± 0%    3.88µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64/useSel=true/hasNulls=true-16              3.14µs ± 0%    3.85µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64Const/useSel=true/hasNulls=true-16         2.44µs ± 0%    3.00µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64/useSel=false/hasNulls=false-16             995ns ± 0%    1303ns ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64Const/useSel=false/hasNulls=false-16        825ns ± 0%    1075ns ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64/useSel=false/hasNulls=true-16             2.26µs ± 0%    2.87µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64Const/useSel=false/hasNulls=true-16        1.62µs ± 0%    1.99µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64/useSel=true/hasNulls=false-16             3.11µs ± 0%    4.57µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64Const/useSel=true/hasNulls=false-16        1.47µs ± 0%    4.08µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64/useSel=true/hasNulls=true-16              2.97µs ± 0%    3.83µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64Const/useSel=true/hasNulls=true-16         2.37µs ± 0%    2.98µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64/useSel=false/hasNulls=false-16            1.00µs ± 0%    1.31µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64Const/useSel=false/hasNulls=false-16        839ns ± 0%     969ns ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64/useSel=false/hasNulls=true-16             2.26µs ± 0%    2.95µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64Const/useSel=false/hasNulls=true-16        1.60µs ± 0%    2.00µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64/useSel=true/hasNulls=false-16             3.06µs ± 0%    4.62µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64Const/useSel=true/hasNulls=false-16        2.40µs ± 0%    4.16µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64/useSel=true/hasNulls=true-16              3.02µs ± 0%    3.82µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64Const/useSel=true/hasNulls=true-16         2.34µs ± 0%    2.97µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64/useSel=false/hasNulls=false-16             963ns ± 0%    1317ns ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64Const/useSel=false/hasNulls=false-16        771ns ± 0%    1110ns ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64/useSel=false/hasNulls=true-16             2.16µs ± 0%    2.97µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64Const/useSel=false/hasNulls=true-16        1.58µs ± 0%    2.04µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64/useSel=true/hasNulls=false-16             2.95µs ± 0%    4.56µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64Const/useSel=true/hasNulls=false-16        2.50µs ± 0%    3.99µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64/useSel=true/hasNulls=true-16              3.01µs ± 0%    3.83µs ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64Const/useSel=true/hasNulls=true-16         2.27µs ± 0%    2.97µs ± 0%   ~     (p=1.000 n=1+1)

name                                                            old speed      new speed      delta
ProjOp/projPlusInt64Int64/useSel=false/hasNulls=false-16        11.0GB/s ± 0%  11.7GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64Const/useSel=false/hasNulls=false-16   6.24GB/s ± 0%  7.86GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64/useSel=false/hasNulls=true-16         4.64GB/s ± 0%  5.38GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64Const/useSel=false/hasNulls=true-16    3.26GB/s ± 0%  3.47GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64/useSel=true/hasNulls=false-16         5.88GB/s ± 0%  5.69GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64Const/useSel=true/hasNulls=false-16    3.50GB/s ± 0%  3.46GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64/useSel=true/hasNulls=true-16          3.36GB/s ± 0%  3.18GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projPlusInt64Int64Const/useSel=true/hasNulls=true-16     2.28GB/s ± 0%  1.92GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64/useSel=false/hasNulls=false-16       12.0GB/s ± 0%  10.3GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64Const/useSel=false/hasNulls=false-16  6.96GB/s ± 0%  7.16GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64/useSel=false/hasNulls=true-16        5.03GB/s ± 0%  4.62GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64Const/useSel=false/hasNulls=true-16   3.21GB/s ± 0%  2.96GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64/useSel=true/hasNulls=false-16        6.63GB/s ± 0%  6.12GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64Const/useSel=true/hasNulls=false-16   3.82GB/s ± 0%  3.70GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64/useSel=true/hasNulls=true-16         3.80GB/s ± 0%  3.96GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMinusInt64Int64Const/useSel=true/hasNulls=true-16    2.41GB/s ± 0%  2.41GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64/useSel=false/hasNulls=false-16        8.95GB/s ± 0%  9.16GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64Const/useSel=false/hasNulls=false-16   5.61GB/s ± 0%  5.86GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64/useSel=false/hasNulls=true-16         4.59GB/s ± 0%  4.70GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64Const/useSel=false/hasNulls=true-16    2.70GB/s ± 0%  2.87GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64/useSel=true/hasNulls=false-16         5.04GB/s ± 0%  5.46GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64Const/useSel=true/hasNulls=false-16    3.11GB/s ± 0%  3.33GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64/useSel=true/hasNulls=true-16          3.52GB/s ± 0%  3.61GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projMultInt64Int64Const/useSel=true/hasNulls=true-16     2.22GB/s ± 0%  2.31GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64/useSel=false/hasNulls=false-16         73.0MB/s ± 0%  79.0MB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64Const/useSel=false/hasNulls=false-16    48.1MB/s ± 0%  49.0MB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64/useSel=false/hasNulls=true-16          91.4MB/s ± 0%  83.2MB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64Const/useSel=false/hasNulls=true-16     48.2MB/s ± 0%  43.6MB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64/useSel=true/hasNulls=false-16          65.4MB/s ± 0%  69.8MB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64Const/useSel=true/hasNulls=false-16     42.1MB/s ± 0%  43.0MB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64/useSel=true/hasNulls=true-16           96.5MB/s ± 0%  77.0MB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projDivInt64Int64Const/useSel=true/hasNulls=true-16      45.2MB/s ± 0%  47.3MB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64/useSel=false/hasNulls=false-16          11.9GB/s ± 0%  10.9GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64Const/useSel=false/hasNulls=false-16     7.21GB/s ± 0%  6.95GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64/useSel=false/hasNulls=true-16           5.48GB/s ± 0%  4.90GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64Const/useSel=false/hasNulls=true-16      3.84GB/s ± 0%  3.75GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64/useSel=true/hasNulls=false-16           3.92GB/s ± 0%  3.35GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64Const/useSel=true/hasNulls=false-16      2.56GB/s ± 0%  1.95GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64/useSel=true/hasNulls=true-16            4.13GB/s ± 0%  4.07GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projEQInt64Int64Const/useSel=true/hasNulls=true-16       2.68GB/s ± 0%  2.59GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64/useSel=false/hasNulls=false-16          12.4GB/s ± 0%  12.1GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64Const/useSel=false/hasNulls=false-16     6.27GB/s ± 0%  7.32GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64/useSel=false/hasNulls=true-16           5.27GB/s ± 0%  5.29GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64Const/useSel=false/hasNulls=true-16      3.48GB/s ± 0%  3.75GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64/useSel=true/hasNulls=false-16           3.66GB/s ± 0%  3.29GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64Const/useSel=true/hasNulls=false-16      2.04GB/s ± 0%  1.86GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64/useSel=true/hasNulls=true-16            3.75GB/s ± 0%  3.97GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projNEInt64Int64Const/useSel=true/hasNulls=true-16       2.68GB/s ± 0%  2.63GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64/useSel=false/hasNulls=false-16          12.2GB/s ± 0%  11.7GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64Const/useSel=false/hasNulls=false-16     7.26GB/s ± 0%  6.56GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64/useSel=false/hasNulls=true-16           5.37GB/s ± 0%  5.09GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64Const/useSel=false/hasNulls=true-16      3.76GB/s ± 0%  3.80GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64/useSel=true/hasNulls=false-16           3.93GB/s ± 0%  3.53GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64Const/useSel=true/hasNulls=false-16      2.68GB/s ± 0%  2.11GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64/useSel=true/hasNulls=true-16            3.99GB/s ± 0%  4.25GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLTInt64Int64Const/useSel=true/hasNulls=true-16       2.57GB/s ± 0%  2.73GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64/useSel=false/hasNulls=false-16          12.6GB/s ± 0%  12.6GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64Const/useSel=false/hasNulls=false-16     7.58GB/s ± 0%  7.62GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64/useSel=false/hasNulls=true-16           5.53GB/s ± 0%  5.72GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64Const/useSel=false/hasNulls=true-16      3.86GB/s ± 0%  4.11GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64/useSel=true/hasNulls=false-16           4.02GB/s ± 0%  3.58GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64Const/useSel=true/hasNulls=false-16      4.25GB/s ± 0%  2.01GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64/useSel=true/hasNulls=true-16            4.21GB/s ± 0%  4.27GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projLEInt64Int64Const/useSel=true/hasNulls=true-16       2.63GB/s ± 0%  2.75GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64/useSel=false/hasNulls=false-16          12.5GB/s ± 0%  12.5GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64Const/useSel=false/hasNulls=false-16     7.46GB/s ± 0%  8.45GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64/useSel=false/hasNulls=true-16           5.53GB/s ± 0%  5.55GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64Const/useSel=false/hasNulls=true-16      3.92GB/s ± 0%  4.09GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64/useSel=true/hasNulls=false-16           4.09GB/s ± 0%  3.54GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64Const/useSel=true/hasNulls=false-16      2.60GB/s ± 0%  1.97GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64/useSel=true/hasNulls=true-16            4.14GB/s ± 0%  4.28GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGTInt64Int64Const/useSel=true/hasNulls=true-16       2.68GB/s ± 0%  2.76GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64/useSel=false/hasNulls=false-16          13.0GB/s ± 0%  12.4GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64Const/useSel=false/hasNulls=false-16     8.11GB/s ± 0%  7.38GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64/useSel=false/hasNulls=true-16           5.79GB/s ± 0%  5.51GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64Const/useSel=false/hasNulls=true-16      3.96GB/s ± 0%  4.02GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64/useSel=true/hasNulls=false-16           4.24GB/s ± 0%  3.60GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64Const/useSel=true/hasNulls=false-16      2.51GB/s ± 0%  2.05GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64/useSel=true/hasNulls=true-16            4.15GB/s ± 0%  4.27GB/s ± 0%   ~     (p=1.000 n=1+1)
ProjOp/projGEInt64Int64Const/useSel=true/hasNulls=true-16       2.75GB/s ± 0%  2.76GB/s ± 0%   ~     (p=1.000 n=1+1)
```

yuzefovich reviewed (commented):
> Reviewed 7 of 7 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84159)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rytaft, and @wenyihu6)

___
*[`-- commits` line 13 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6t3gq85YMIZ41LAdUd):*
nit: I think we should be more specific since AFAIK only Concat operation on arrays is susceptible to this.
___
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 125 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6t5J_vBlRRd7o6sSV8:-N6t5J_vBlRRd7o6sSV9:b-rztigr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7facf1eb091b8dbe085437a9f2fe9f486b397d/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L125)):*
> ```Go
> 		_outNulls := projVec.Nulls()
> 
> 		// If nullableArgs is true, the function’s definition can handle null
> ```

Let's make these comments "template-time only" (i.e. exclude them from the generated code so that it is easier to see how actual generated code is changed).

We can do so with the following syntax:
```
// {{/*
//        TEMPLATE TIME ONLY COMMENT
// */}}
```
meaning that `TEMPLATE TIME ONLY COMMENT` line is only visible in `_tmpl.go` file but is omitted from the `eg.go` file.

To get there, first let's add another commit that makes already present comments template-time only:
```
--- a/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go
+++ b/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go
@@ -115,10 +115,12 @@ func (p _OP_NAME) Next() coldata.Batch {
                vec2 := batch.ColVec(p.col2Idx)
                col1 := vec1._L_TYP()
                col2 := vec2._R_TYP()
-               // Some operators can result in NULL with non-NULL inputs, like the JSON
-               // fetch value operator, ->. Therefore, _outNulls is defined to allow
-               // updating the output Nulls from within _ASSIGN functions when the result
-               // of a projection is Null.
+               // {{/*
+               //     Some operators can result in NULL with non-NULL inputs, like the
+               //     JSON fetch value operator, ->. Therefore, _outNulls is defined to
+               //     allow updating the output Nulls from within _ASSIGN functions
+               //     when the result of a projection is Null.
+               // */}}
                _outNulls := projVec.Nulls()
                if vec1.Nulls().MaybeHasNulls() || vec2.Nulls().MaybeHasNulls() {
                        _SET_PROJECTION(true)
@@ -154,11 +156,13 @@ func _SET_PROJECTION(_HAS_NULLS bool) {
                        _SET_SINGLE_TUPLE_PROJECTION(_HAS_NULLS, false)
                }
        }
-       // _outNulls has been updated from within the _ASSIGN function to include
-       // any NULLs that resulted from the projection.
-       // If _HAS_NULLS is true, union _outNulls with the set of input Nulls.
-       // If _HAS_NULLS is false, then there are no input Nulls. _outNulls is
-       // projVec.Nulls() so there is no need to call projVec.SetNulls().
+       // {{/*
+       //     _outNulls has been updated from within the _ASSIGN function to
+       //     include any NULLs that resulted from the projection.
+       //     If _HAS_NULLS is true, union _outNulls with the set of input Nulls.
+       //     If _HAS_NULLS is false, then there are no input Nulls. _outNulls is
+       //     projVec.Nulls() so there is no need to call projVec.SetNulls().
+       // */}}
        // {{if _HAS_NULLS}}
        projVec.SetNulls(_outNulls.Or(*col1Nulls).Or(*col2Nulls))
        // {{end}}
```
(and the same for `pkg/sql/colexec/colexecprojconst/proj_const_ops_tmpl.go`). These changes would be in the separate first commit.

Then, all the new comments you're adding would be included into the second commit with the bug fix and also template-time only.
___
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 132 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6t6Vvd2AzQgsgyZIfI:-N6t6Vvd2AzQgsgyZIfJ:b-ubqnif) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7facf1eb091b8dbe085437a9f2fe9f486b397d/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L132)):*
> ```Go
> 		// another projection operation that has nullableArgs == true, we should
> 		// update this code accordingly.
> 		// {{if and (eq .Left.VecMethod "Datum") (eq .Right.VecMethod "Datum") (eq .Name "Concat")}}
> ```

Maybe - to make things more bullet-proof - let's use the conditional such that if either is `Datum`, regardless of the `Name`, we'd look at `p.nullableArgs`? WDYT?
___
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 172 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6t64c02ELnNld-LCMg:-N6t64c02ELnNld-LCMh:bxj1nt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7facf1eb091b8dbe085437a9f2fe9f486b397d/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L172)):*
> ```Go
> 	}
> 	// _outNulls has been updated from within the _ASSIGN function to include any
> 	// NULLs that resulted from the projection. 
> ```

nit: there is unnecessary change in this line.
___
*[`pkg/sql/colexec/colexecprojconst/proj_const_ops_tmpl.go` line 138 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6t6OGQAJbYFSu_1agF:-N6t6OGQAJbYFSu_1agG:bst6t6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7facf1eb091b8dbe085437a9f2fe9f486b397d/pkg/sql/colexec/colexecprojconst/proj_const_ops_tmpl.go#L138)):*
> ```Go
> 		// another projection operation that has nullableArgs == true, we should
> 		// update this code accordingly.
> 		// {{if or (eq .Left.VecMethod "Datum") (eq .Right.VecMethod "Datum")}}
> ```

nit: why have a different conditional structure here than in `proj_non_const_ops_tmpl.go`?


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 125 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6t5J_vBlRRd7o6sSV8:-N6xncZY3TOVnY30166m:bm67l3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7facf1eb091b8dbe085437a9f2fe9f486b397d/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L125)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Let's make these comments "template-time only" (i.e. exclude them from the generated code so that it is easier to see how actual generated code is changed).

We can do so with the following syntax:
```
// {{/*
//        TEMPLATE TIME ONLY COMMENT
// */}}
```
meaning that `TEMPLATE TIME ONLY COMMENT` line is only visible in `_tmpl.go` file but is omitted from the `eg.go` file.

To get there, first let's add another commit that makes already present comments template-time only:
```
--- a/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go
+++ b/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go
@@ -115,10 +115,12 @@ func (p _OP_NAME) Next() coldata.Batch {
                vec2 := batch.ColVec(p.col2Idx)
                col1 := vec1._L_TYP()
                col2 := vec2._R_TYP()
-               // Some operators can result in NULL with non-NULL inputs, like the JSON
-               // fetch value operator, ->. Therefore, _outNulls is defined to allow
-               // updating the output Nulls from within _ASSIGN functions when the result
-               // of a projection is Null.
+               // {{/*
+               //     Some operators can result in NULL with non-NULL inputs, like the
+               //     JSON fetch value operator, ->. Therefore, _outNulls is defined to
+               //     allow updating the output Nulls from within _ASSIGN functions
+               //     when the result of a projection is Null.
+               // */}}
                _outNulls := projVec.Nulls()
                if vec1.Nulls().MaybeHasNulls() || vec2.Nulls().MaybeHasNulls() {
                        _SET_PROJECTION(true)
@@ -154,11 +156,13 @@ func _SET_PROJECTION(_HAS_NULLS bool) {
                        _SET_SINGLE_TUPLE_PROJECTION(_HAS_NULLS, false)
                }
        }
-       // _outNulls has been updated from within the _ASSIGN function to include
-       // any NULLs that resulted from the projection.
-       // If _HAS_NULLS is true, union _outNulls with the set of input Nulls.
-       // If _HAS_NULLS is false, then there are no input Nulls. _outNulls is
-       // projVec.Nulls() so there is no need to call projVec.SetNulls().
+       // {{/*
+       //     _outNulls has been updated from within the _ASSIGN function to
+       //     include any NULLs that resulted from the projection.
+       //     If _HAS_NULLS is true, union _outNulls with the set of input Nulls.
+       //     If _HAS_NULLS is false, then there are no input Nulls. _outNulls is
+       //     projVec.Nulls() so there is no need to call projVec.SetNulls().
+       // */}}
        // {{if _HAS_NULLS}}
        projVec.SetNulls(_outNulls.Or(*col1Nulls).Or(*col2Nulls))
        // {{end}}
```
(and the same for `pkg/sql/colexec/colexecprojconst/proj_const_ops_tmpl.go`). These changes would be in the separate first commit.

Then, all the new comments you're adding would be included into the second commit with the bug fix and also template-time only.
</blockquote></details>

Sounds great! I will make the changes soon. Which part of the bug fix do you want me to include in the second commit? The bug fix is part of the first commit. 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84159)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rytaft, and @wenyihu6)

___
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 125 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6t5J_vBlRRd7o6sSV8:-N6xp3Ar0US_IzLQjfIP:b-z9turf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7facf1eb091b8dbe085437a9f2fe9f486b397d/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L125)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Sounds great! I will make the changes soon. Which part of the bug fix do you want me to include in the second commit? The bug fix is part of the first commit. 
</blockquote></details>

I was thinking that you would change the order of the commits: the first one would be the suggestion to make the existing comments template-time only, and then the second commit would be what is currently included in this PR.

It doesn't have to be this way - feel free to keep your current bug-fix commit as the first one (while making the changes to make all new comments template-time only) and then add a second commit to make the pre-existing comments template-time only. Up to you which way to go.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 132 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6t6Vvd2AzQgsgyZIfI:-N6xpanE93HWODoJdE7Y:baqu3l3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7facf1eb091b8dbe085437a9f2fe9f486b397d/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L132)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Maybe - to make things more bullet-proof - let's use the conditional such that if either is `Datum`, regardless of the `Name`, we'd look at `p.nullableArgs`? WDYT?
</blockquote></details>

There are a couple of other functions with Datum as its left's or right's `VecMethod`. For example, `projBitandDatumDatumOp`, , `projPlusInt32DatumOp`. `p.nullableArgs` in the if statements would be not be used. We could use if both are `Datum`, and there will be less functions. They are functions like `projBitandDatumDatumOp`, `projBitandDatumDatumOp`, `projBitorDatumDatumOp`. But there aren't too many, so the performance different is probably negligible. 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84159)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rytaft, and @wenyihu6)

___
*[`pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go` line 132 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N6t6Vvd2AzQgsgyZIfI:-N6xtHpO9Pr3Z_DgusyN:b-p80rq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7facf1eb091b8dbe085437a9f2fe9f486b397d/pkg/sql/colexec/colexecproj/proj_non_const_ops_tmpl.go#L132)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

There are a couple of other functions with Datum as its left's or right's `VecMethod`. For example, `projBitandDatumDatumOp`, , `projPlusInt32DatumOp`. `p.nullableArgs` in the if statements would be not be used. We could use if both are `Datum`, and there will be less functions. They are functions like `projBitandDatumDatumOp`, `projBitandDatumDatumOp`, `projBitorDatumDatumOp`. But there aren't too many, so the performance different is probably negligible. 
</blockquote></details>

Yeah, that sounds good - the performance impact should be minor, but the code will be less error-prone.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
colexec: changed comments to template-time only

This commit changed some comments in `proj_const_ops_tmpl.go` and
`proj_non_const_ops_tmpl.go` to become template-time only.

Release note: none

yuzefovich reviewed (approved):
Nice work! I only have some nits about the release note, but otherwise <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 11 of 11 files at r8, 11 of 11 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-:-N7HRgxp8jdfNJhl3LwI:b-9vooh1)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @rytaft, and @wenyihu6)

___
*[`-- commits` line 26 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N7HQAaF-Jcv4pUC4Prz):*
nit: I think we should mention Concat operator on arrays explicitly since this is the only known reproduction of the bug. Without the mention, the bug might sound much worse than it actually is.
___
*[`-- commits` line 27 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-N7HPjZ166gsb6Xk8PWA):*
nit: these `Release note`s are ultimately made for user-consumption, so the second sentence should be removed (users don't need to know about such low level details).


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
Does it look better now?





<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Yep, perfect, thanks!

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84159#-:-N7HVK2w8kpm7_PSKcrr:bu0rvje)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @rytaft)



<!-- Sent from Reviewable.io -->

wenyihu6 commented:
bors r+

rytaft commented:
Thanks for fixing this! Can we backport it to 22.1 and 21.2?





<!-- Sent from Reviewable.io -->

yuzefovich commented:
> Can we backport it to 22.1 and 21.2?

Sure. @wenyihu6 let me know if you need some help with that. You can start with `backport -r 22.1 84159` using our [backport](https://github.com/benesch/backport) tool (that you'll need to build manually); however, that will run into merge conflicts which will need to be resolved.
